### PR TITLE
Support xdrip-js Sensor State Pill

### PR DESCRIPTION
**Pull request by jpcunningh**
Adds new pill type to support displaying glucose sensor state.

### CODE REVIEW

**Comment by sulkaharo** [comment]
@jpcunningh Can you confirm this is ready for merging? :)

**Comment by jpcunningh** [comment]
Not yet, we are still testing it to ensure it effectively supports both Lookout and Logger.

I'll stick a note here when we are comfortable with it.

Thanks, @sulkaharo!

**Comment by sulkaharo** [comment]
We need to publish a new release in less than 10 days, so if you want this to go into next Master, make it quick. :) (mLab is pushing out an update that'll break Master on 20th)

**Comment by jpcunningh** [comment]
We'll make it!

What day would you need to merge?

Thanks!

**Comment by PieterGit** [comment]
If ok it can be added to the 0.10.3 release. I can't test xdrip, so others should 👍  it.

**Comment by efidoman** [comment]
I'm testing this one now and will update with any findings.

File: lib/plugins/xdrip-js.js [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+var levels = require('../levels');
+
+function init(ctx) {
+  var utils = require('../utils')(ctx);
+  var firstPrefs = true;
+  var lastStateNotification = null;
+
+  var sensorState = {
+    name: 'xdrip-js'
+    , label: 'Sensor Status'
```
Comment (efidoman) [inline lib/plugins/xdrip-js.js]: Just a suggestion ... should we consider renaming this to something more broad like "CGM Status"?

**Comment by jpcunningh** [comment]
I agree, CGM Status is better. I'll change it this evening.

Are you good with the plugin name xdrip-js? I thought about a more general
name that could be used by other applications such as xdrip+, but that may
be more constraining than letting them copy the code and update it to match
their needs.

File: lib/plugins/xdrip-js.js (hunk 1 above)
Comment (efidoman) [reply lib/plugins/xdrip-js.js]: Sounds good ...
Is ```cgm``` already used

**Comment by efidoman** [comment]
Testing tonight with Logger. Looks great so far!
7-9-2018 5-07-49 pm

**Comment by efidoman** [comment]
More Testing this morning. Working great.
cgm-pill

**Comment by jpcunningh** [comment]
@efidoman, I'll be out of communication for the day. I made the change a moment ago to only print session age if valid data is available instead of print N/A when no valid data was available.

I'm happy with it as it is for merging.  If you think it's ready to merge, please post here to let @sulkaharo know so we can make this release.

Thanks!

Jeremy

**Comment by efidoman** [comment]
@sulkaharo - My testing of this branch with Logger is complete. I'm good with the changes in this branch.

**Comment by PieterGit** [comment]
I scanned the code change and because both @efidoman @jpcunningh I decided to merge this to dev. Thanks for the PR

**Comment by jpcunningh** [comment]
It is xdrip-js.  Other apps can use it as is if it works or copy and modify
as needed. If nothing else,  it should be a good starting point for xdrip,
spike, etc.

**Comment by jaylagorio** [comment]
I just updated an installation on `dev` with this PR and included `xdrip-js` at the end of the `ENABLE` string:

`careportal iob cob openaps pump bwg rawbg basal xdrip-js`

It doesn't show up automatically and I don't see a checkbox in the options menu to show it. What did I miss?

**Comment by jpcunningh** [comment]
I'm not sure. I have to refresh my browser a couple of times with the developer tools open (I have the setting to disable browser caching if the developer tools are open) to get it to refresh the client side code.

My enable string is:
`openaps pump iob careportal sage cage maker upbat rawbg pushover xdrip-js`

I also have `DEVICESTATUS_ADVANCED` set to true.

**Comment by UphwbKpqSm47z3qewH9x** [comment]
Enabled the plugin xdrip-js, pill shows up but I've got unknown for all fields.
image

verified DEVICESTATUS_ADVANCED = true
if I peak at my mlab devicestatus collection, I only see entries for my phone's battery as uploader and no CGM stats.
image

G5 collector on xdrip+ nightly, REST-API uploader

any hints on what I might be missing? Config in xdrip or nightscout?

**Comment by PieterGit** [comment]
@UphwbKpqSm47z3qewH9x can't help you with this one. Please discuss issues with this PR in a seperate issue. I'm hoping @jpcunningh can help you.

**Comment by jaylagorio** [comment]
@jpcunningh Not sure what happened, but now the CGM Status checkbox is there. I checked it, refreshed and... everything is set to unknown. I'm running Logger if that helps. Any thoughts?

**Comment by jpcunningh** [comment]
Let's move over to the xdrip gitter channel:
https://gitter.im/thebookins/xdrip-js

**Comment by efidoman** [comment]
The initial Logger support the CGM pill  has been merged to xdrip-js/Logger dev. I am still going to work on enhancing it, but the basic functionality is very nice to have. You will need to update Logger dev to latest ```cd ~/src/Logger && git pull && npm run global-install``` and deploy (redeploy) the NS dev branch (in heroku or azure). You will need to set the NS ENABLE flag adding xdrip-js. One word of caution... I strongly recommend waiting to update Logger when you are doing a new sensor insert. Otherwise, you might have to copy over or restart calibrations.

**Pull request merged**